File: java/src/main/java/com/taobao/top/link/endpoint/Endpoint.java
package com.taobao.top.link.endpoint;

import java.net.URI;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;

import com.taobao.top.link.DefaultLoggerFactory;
import com.taobao.top.link.LinkException;
import com.taobao.top.link.Logger;
import com.taobao.top.link.LoggerFactory;
import com.taobao.top.link.channel.ChannelException;
import com.taobao.top.link.channel.ChannelSender;
import com.taobao.top.link.channel.ClientChannel;
import com.taobao.top.link.channel.ClientChannelSelector;
import com.taobao.top.link.channel.ServerChannel;

// Abstract network model
// https://docs.google.com/drawings/d/1PRfzMVNGE4NKkpD9A_-QlH2PV47MFumZX8LbCwhzpQg/edit
public class Endpoint {
	protected static int TIMOUTSECOND = 5;
	private Logger logger;
	private Identity identity;
	private List<ServerChannel> serverChannels;
	private ClientChannelSelector channelSelector;
	private EndpointChannelHandler channelHandler;
	private MessageHandler messageHandler;

	// in/out endpoints
	private List<EndpointProxy> connected;

	public Endpoint(Identity identity) {
		this(new DefaultLoggerFactory(), identity);
	}

	public Endpoint(LoggerFactory loggerFactory, Identity identity) {
		this.serverChannels = new ArrayList<ServerChannel>();
		this.connected = new ArrayList<EndpointProxy>();
		this.logger = loggerFactory.create(this);
		this.identity = identity;
		this.channelSelector = new ClientChannelSharedSelector(loggerFactory);
		this.channelHandler = new EndpointChannelHandler(loggerFactory, this);

		if (this.identity == null)
			throw new NullPointerException("identity");
	}

	public Identity getIdentity() {
		return this.identity;
	}

	public void setMessageHandler(MessageHandler handler) {
		this.messageHandler = handler;
	}

	public MessageHandler getMessageHandler() {
		return this.messageHandler;
	}

	public void bind(ServerChannel channel) {
		channel.setChannelHandler(this.channelHandler);
		channel.run();
		this.serverChannels.add(channel);
	}

	public void unbindAll() {
		for (ServerChannel channel : this.serverChannels) {
			try {
				channel.stop();
			} catch (Exception e) {
				this.logger.error("unbind error", e);
			}
		}
		this.serverChannels.clear();
	}

	public Iterator<EndpointProxy> getConnected() {
		return this.connected.iterator();
	}

	public synchronized EndpointProxy getEndpoint(Identity targetIdentity, URI uri) throws LinkException {
		EndpointProxy e = this.getEndpoint(targetIdentity);
		// always clear, cached proxy will have broken channel
		e.remove(uri);
		// always reget channel, make sure it's valid
		ClientChannel channel = this.channelSelector.getChannel(uri);
		channel.setChannelHandler(this.channelHandler);
		e.add(channel);
		// connect message
		Message msg = new Message();
		msg.messageType = MessageType.CONNECT;
		HashMap<String, String> content = new HashMap<String, String>();
		this.identity.render(content);
		msg.content = content;
		this.sendAndWait(e, channel, msg, TIMOUTSECOND);
		return e;
	}

	public synchronized EndpointProxy getEndpoint(Identity identity) throws LinkException {
		if (identity.equals(this.identity))
			throw new LinkException("target identity can not equal itself");

		for (EndpointProxy e : this.connected) {
			if (e.getIdentity() != null &&
					e.getIdentity().equals(identity))
				return e;
		}
		EndpointProxy e = this.createProxy("by identity|" + identity.toString());
		e.setIdentity(identity);
		return e;
	}

	protected void send(ChannelSender sender, Message message) throws ChannelException {
		this.channelHandler.pending(message, sender);
	}

	protected HashMap<String, String> sendAndWait(EndpointProxy e,
			ChannelSender sender,
			Message message,
			int timeoutSecond) throws LinkException {
		SendCallback callback = new SendCallback(e);
		this.channelHandler.pending(message, sender, callback);
		callback.waitReturn(timeoutSecond);
		if (callback.getError() != null) {
			throw callback.getError();
		}
		return callback.getReturn();
	}

	private EndpointProxy createProxy(String reason) {
		EndpointProxy e = new EndpointProxy(this);
		this.connected.add(e);
		if (this.logger.isDebugEnable())
			this.logger.debug("create new EndpointProxy: " + reason);
		return e;
	}
}

File: csharp/src/Taobao.Top.Link/Endpoints/Endpoint.cs
﻿using System;
using System.Collections.Generic;
using System.Text;
using Taobao.Top.Link.Channel;

namespace Taobao.Top.Link.Endpoints
{
    // Abstract network model
    // https://docs.google.com/drawings/d/1PRfzMVNGE4NKkpD9A_-QlH2PV47MFumZX8LbCwhzpQg/edit
    public sealed class Endpoint
    {
        internal static int TIMOUT = 5000;
        private ILog _log;
        private IClientChannelSelector _channelSelector;
        // in/out endpoints
        private IList<EndpointProxy> _connected;
        private EndpointHandler _handler;

        /// <summary>message received, see RTT based
        /// </summary>
        public event EventHandler<EndpointContext> OnMessage;
        /// <summary>ack message received, see RTT based
        /// </summary>
        public event EventHandler<AckMessageArgs> OnAckMessage;

        /// <summary>get id
        /// </summary>
        public Identity Identity { get; private set; }

        public Endpoint(Identity identity) : this(DefaultLoggerFactory.Default, identity) { }
        public Endpoint(ILoggerFactory loggerFactory, Identity identity)
        {
            this._connected = new List<EndpointProxy>();
            this._log = loggerFactory.Create(this);
            this.Identity = identity;
            this._channelSelector = new ClientChannelSharedSelector(loggerFactory);
            this._handler = new EndpointHandler(loggerFactory);
            this._handler.MessageHandler = ctx => OnMessage(this, ctx);
            this._handler.AckMessageHandler = (m, i) => OnAckMessage(this, new AckMessageArgs(m, i));
        }

        /// <summary>get connected endpoint by id
        /// </summary>
        /// <param name="target"></param>
        /// <returns></returns>
        public EndpointProxy GetEndpoint(Identity target)
        {
            if (target.Equals(this.Identity))
                throw new LinkException(Text.E_ID_DUPLICATE);
            foreach (EndpointProxy p in this._connected)
                if (p.Identity != null && p.Identity.Equals(target))
                    return p;
            return null;
        }
        /// <summary>connect endpoint
        /// </summary>
        /// <param name="target">target id</param>
        /// <param name="uri">target address</param>
        /// <returns></returns>
        public EndpointProxy GetEndpoint(Identity target, string uri)
        {
            return this.GetEndpoint(target, uri, null);
        }
        /// <summary>connect endpoint
        /// </summary>
        /// <param name="target">target id</param>
        /// <param name="uri">target address</param>
        /// <param name="extras">passed as connect message</param>
        /// <returns></returns>
        public EndpointProxy GetEndpoint(Identity target, string uri, IDictionary<string, string> extras)
        {
            EndpointProxy e = this.GetEndpoint(target) ?? this.CreateProxy();
            e.Identity = target;
            // always clear, cached proxy will have broken channel
            e.Remove(uri);
            // always reget channel, make sure it's valid
            IClientChannel channel = this._channelSelector.GetChannel(new Uri(uri));
            e.Add(channel);
            // connect message
            Message msg = new Message();
            msg.MessageType = MessageType.CONNECT;
            IDictionary<string, string> content = new Dictionary<string, string>();
            this.Identity.Render(content);
            // pass extra data
            if (extras != null)
                foreach (var p in extras)
                    content.Add(p);
            msg.Content = content;
            this._handler.SendAndWait(e, channel, msg, TIMOUT);
            return e;
        }

        private EndpointProxy CreateProxy()
        {
            EndpointProxy e = new EndpointProxy(this._handler);
            this._connected.Add(e);
            return e;
        }
    }
}